core/src/main/java/hudson/PluginManager.java

```diff
@@ -778,45 +944,117 @@ public HttpResponse doUpdateSources(StaplerRequest req) throws IOException {
      * Performs the installation of the plugins.
      */
     public void doInstall(StaplerRequest req, StaplerResponse rsp) throws IOException, ServletException {
-        boolean dynamicLoad = req.getParameter("dynamicLoad")!=null;
+        Set<String> plugins = new LinkedHashSet<>();
 
         Enumeration<String> en = req.getParameterNames();
         while (en.hasMoreElements()) {
             String n =  en.nextElement();
             if(n.startsWith("plugin.")) {
                 n = n.substring(7);
-                // JENKINS-22080 plugin names can contain '.' as could (according to rumour) update sites
-                int index = n.indexOf('.');
-                UpdateSite.Plugin p = null;
+                plugins.add(n);
+            }
+        }
+
+        boolean dynamicLoad = req.getParameter("dynamicLoad")!=null;
+        install(plugins, dynamicLoad);
+
+        rsp.sendRedirect("../updateCenter/");
+    }
+
+    /**
+     * Installs a list of plugins from a JSON POST.
+     * @param req The request object.
+     * @return A JSON response that includes a "correlationId" in the "data" element.
+     * That "correlationId" can then be used in calls to
+     * {@link UpdateCenter#doInstallStatus(org.kohsuke.stapler.StaplerRequest)}.
+     * @throws IOException Error reading JSON payload fro request.
+     * @since FIXME
+     */
+    @RequirePOST
+    public HttpResponse doInstallPlugins(StaplerRequest req) throws IOException {
```

Comment:
Ok, we're getting subtle :smile: (but it is a good point).

The thing is .... I see `doThis` and `doThat` methods all over Jenkins (all web methods) and it was never a pattern until now (or I just haven't been aware of it perhaps). I'm not complaining though coz I do think it makes sense (I'm sure there would be some exceptions of course) to restrict these methods and not have Java based calls to them. The ones that take a `StaplerRequest` (and maybe return a `HttpResponse`) are 99.9% guaranteed to fit this.